### PR TITLE
Filter is_purchasable when getting cart item from session

### PR DESCRIPTION
**Pull request by helgatheviking**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Introduce `woocommerce_cart_item_is_purchasable ` filter and provide access to the cart item data array. 

### How to test the changes in this Pull Request:

1. add this snippet to `functions.php`
```
function kia_filter_product_in_cart( $is_purchasable, $cart_key, $cart_item_data ) {
    if( $is_purchasable && isset( $cart_item_data['testing'] ) ) {
    	$is_purchasable = false;
    }
    return $is_purchasable;
}
add_filter( 'woocommerce_cart_item_is_purchasable', 'kia_filter_product_in_cart', 10, 3 );


function kia_add_test_item_data( $cart_item_data ) {
    $cart_item_data['testing'] = 1;
    return $cart_item_data;
}
add_filter( 'woocommerce_add_cart_item', 'kia_add_test_item_data' );
```
2. Add a simple product to the cart
3. Go to cart, the product should have been removed. 

This isn't a very practical test, but it shows it works. The use-case I'm envisioning is that Mix and Match and Bundles sometimes need to sell products _only_ as part of the bundle... ie: not sold separately. It's already possible to manipulate the purchasable status inside the bundle and both Mix and Match and Bundles add the bundled products to the cart, bypassing the default Woo process. However, when the cart is loaded from the session we hit this roadblock. Without access to the cart item data, there's nothing to use to filter `is_purchasable()` and report that this product is actually purchasable when being ordered as part of a bundle.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> New: `woocommerce_cart_item_is_purchasable ` filter

cc: @franticpsyx

### CODE REVIEW

File: includes/class-wc-cart-session.php [hunk 1]
```diff
@@ -128,7 +128,16 @@ public function get_cart_from_session() {
...
+			/**
+			 * Allow 3rd parties to override this item's is_puchasable() result with cart item data.
+			 *
+			 * @since 4.1.0
+			 *
+			 * @param bool $is_purchasable If false, the item will not be added to the cart. Default: product's is_purchasable() status.
+			 * @param string $key Cart item key.
+			 * @param array $values Cart item values e.g. quantity and product_id.
+			 */
+			} elseif ( ! apply_filters( 'woocommerce_cart_item_is_purchasable', $product->is_purchasable(), $key, $values ) ) {
```
Comment (claudiosanches) [inline includes/class-wc-cart-session.php:140]: You can achieve it already with `woocommerce_pre_remove_cart_item_from_session`.
If you like we could change to `apply_filters( 'woocommerce_pre_remove_cart_item_from_session', false, $key, $values, $product )` so you don't need to load the product data again.

**Review by claudiosanches — CHANGES_REQUESTED** [review]

File: includes/class-wc-cart-session.php (hunk 1 above)
Comment (helgatheviking) [reply includes/class-wc-cart-session.php:140]: Hi @claudiosanches that might work...

Comment (helgatheviking) [reply includes/class-wc-cart-session.php:140]: Though if we're bypassing the `is_purchasable()` check and overriding it's value to keep the product in the cart then we don't need to set `$update_cart_session = true;` if the product isn't getting removed.

Comment (claudiosanches) [reply includes/class-wc-cart-session.php:140]: In both places will set `$update_cart_session`, but with the pre filter you can include any notification you like.

Comment (helgatheviking) [reply includes/class-wc-cart-session.php:140]: If the product is not removed though, why would the cart session need to update? I'm trying to bypass the removal and leave the product in the cart (if it's part of a bundle). Though I do need to think if there's another way to do this.

Comment (claudiosanches) [reply includes/class-wc-cart-session.php:140]: > If the product is not removed though, why would the cart session need to update?

By default it's there, but if it's a problem, just fine to customize it.
`woocommerce_pre_remove_cart_item_from_session` is the way to customize this method, if necessary we can improve it.

Comment (claudiosanches) [reply includes/class-wc-cart-session.php:140]: @helgatheviking do you think if we pass `$product` in the `woocommerce_pre_remove_cart_item_from_session` filter will work for you?

**Comment by peterfabian** [comment]
Hi @helgatheviking 👋 

I'd like to follow up here (and sorry for the delay). Have you found a good solution for your problem and we can close this PR or would you still like us to add this filter? Thanks!

**Review by jorgeatorres — DISMISSED** [review]
LGTM. Thanks @helgatheviking & @vedanshujain!

**Review by jorgeatorres — APPROVED** [review]
:shipit:

**Pull request merged**